File: java/array_search/UniformBinarySearch.java
public class UniformBinarySearch {
    public static int uniformBinarySearch(int[] array, int target) {
        if (array.length == 0) return -1;

        int largest = 1;
        while (largest <= array.length / 2) largest *= 2;

        int base = -1;
        for (int step = largest; step >= 1; step /= 2) {
            int probe = base + step;
            if (probe >= array.length) continue;

            if (array[probe] == target) return probe;
            if (array[probe] < target) base = probe;
        }

        return -1;
    }
}

File: cpp/array_search/uniform_binary_search.cpp
#include <cstddef>
#include <vector>

int uniform_binary_search(const std::vector<int>& array, int target) {
    const std::ptrdiff_t length = static_cast<std::ptrdiff_t>(array.size());
    if (length == 0) return -1;

    std::ptrdiff_t largest = 1;
    while (largest <= length / 2) largest *= 2;

    std::ptrdiff_t base = -1;
    for (std::ptrdiff_t step = largest; step >= 1; step /= 2) {
        const std::ptrdiff_t probe = base + step;
        if (probe >= length) continue;

        if (array[probe] == target) return static_cast<int>(probe);
        if (array[probe] < target) base = probe;

        if (step == 1) break;
    }

    return -1;
}
